File: matiasmasso/Mat.NET/DT2/DTOEdiversaInvoicD96a.vb
Public Class DTOEdiversaInvoicD96a
    Inherits DTOBaseGuid

    Property inv As String
    Property dtm As Date
    Property rffDq As String
    Property rffOn As String
    Property rffVn As String
    Property nadSu As String
    Property nadBy As String
    Property nadIv As String
    Property nadDp As String
    Property cur As String
    Property moaresNet As Decimal
    Property moaresBrut As Decimal
    Property moaresBase As Decimal
    Property moaresLiq As Decimal
    Property moaresTax As Decimal
    Property MoaresDto As Decimal
    Property moaresCharges As Decimal
    Property items As List(Of Item)

    Public Class Item
        Inherits DTOBaseGuid

        Property lin As Integer
        Property ean As String
        Property piaLin As String
        Property imdLin As String
        Property qtyLin As Integer
        Property priLinNet As Decimal
        Property priLinBrut As Decimal
        Property moaLinNet As Decimal

        Public Sub New()
            MyBase.New
        End Sub

        Public Sub New(oGuid As Guid)
            MyBase.New(oGuid)
        End Sub
    End Class

    Public Sub New()
        MyBase.New
        _items = New List(Of Item)
    End Sub

    Public Sub New(oGuid As Guid)
        MyBase.New(oGuid)
        _items = New List(Of Item)
    End Sub

    Shared Function Factory(src As String) As DTOEdiversaInvoicD96a
        Dim segments = src.Split(vbCrLf).ToList
        Dim retval As New DTOEdiversaInvoicD96a
        For Each segment In segments
            Dim fields = segment.Split("|").ToList
            Dim item As New Item
            If fields.Count > 1 Then
                Select Case fields.First
                    Case "INV"
                        retval.inv = Field(segment, 1)
                    Case "DTM"
                        retval.dtm = Field(segment, 1)
                    Case "RFF"
                        Select Case Field(segment, 1)
                            Case "DQ"
                                retval.rffDq = Field(segment, 2)
                            Case "ON"
                                retval.rffOn = Field(segment, 2)
                            Case "VN"
                                retval.rffVn = Field(segment, 2)
                        End Select
                    Case "NADSU"
                        retval.nadSu = Field(segment, 2)
                    Case "NADBY"
                        retval.nadBy = Field(segment, 2)
                    Case "NADIV"
                        retval.nadIv = Field(segment, 2)
                    Case "NADDP"
                        retval.nadDp = Field(segment, 2)
                    Case "CUX"
                        retval.Cur = Field(segment, 2)
                    Case "ALC"
                    Case "MOARES"
                        retval.moaresNet = parseDecimal(segment, 1)
                        retval.moaresBrut = parseDecimal(segment, 2)
                        retval.moaresBase = parseDecimal(segment, 3)
                        retval.moaresLiq = parseDecimal(segment, 4, 10)
                        retval.moaresTax = parseDecimal(segment, 5)
                        retval.MoaresDto = parseDecimal(segment, 6, 13)
                        retval.moaresCharges = parseDecimal(segment, 9)

                    Case "LIN"
                        item = New Item
                        item.ean = Field(segment, 1)
                        item.lin = parseInt(segment, 3)
                    Case "PIALIN"
                        item.piaLin = Field(segment, 1)
                    Case "IMDLIN"
                        item.imdLin = Field(segment, 1)
                    Case "QTYLIN"
                        item.qtyLin = parseInt(segment, 2)
                    Case "MOALIN"
                        item.moaLinNet = parseDecimal(segment, 3)
                    Case "PRILIN"
                        Select Case Field(segment, 2)
                            Case "AAA"
                                item.priLinNet = parseDecimal(segment, 2)
                            Case "AAB"
                                item.priLinBrut = parseDecimal(segment, 2)
                        End Select
                End Select
            End If
        Next
        Return retval
    End Function

    Shared Function parseInt(segment As String, fieldIdx As Integer, Optional alternativeIdx As Integer = 0) As Integer
        Dim retval As Integer = 0
        Dim value As String = Field(segment, fieldIdx, alternativeIdx)
        If TextHelper.VbIsNumeric(value) Then
            retval = value
        End If
        Return retval
    End Function

    Shared Function parseDecimal(segment As String, fieldIdx As Integer, Optional alternativeIdx As Integer = 0) As Decimal
        Dim retval As Integer = 0
        Dim value As String = Field(segment, fieldIdx, alternativeIdx)
        Try
            retval = Decimal.Parse(value, System.Globalization.CultureInfo.InvariantCulture)
        Catch ex As Exception
        End Try
        Return retval
    End Function

    Shared Function ParseFch(segment As String, fieldIdx As Integer, Optional alternativeIdx As Integer = 0) As Date
        Dim retval As Date = Nothing
        Dim value As String = Field(segment, fieldIdx, alternativeIdx)
        Dim sFormat As String = ""
        Dim provider As Globalization.CultureInfo = Globalization.CultureInfo.InvariantCulture
        Try
            Select Case value.Length
                Case 0
                Case 6
                    retval = Date.ParseExact(value, "yyMMdd", provider)
                Case 8
                    retval = Date.ParseExact(value, "yyyyMMdd", provider)
                Case 12
                    retval = Date.ParseExact(value, "yyyyMMddHHmm", provider)
                Case Else
                    'exs.Add(DTOEdiversaException.Factory(DTOEdiversaException.Cods.BadFormatFch, src, "No es pot convertir " & src & " en una data"))
            End Select

        Catch ex As Exception
            'exs.Add(DTOEdiversaException.Factory(DTOEdiversaException.Cods.BadFormatFch, src))
        End Try

        Return retval
    End Function

    Shared Function Field(segment As String, fieldIdx As Integer, Optional alternativeIdx As Integer = 0) As String
        Dim retval As String = ""
        Dim fields = segment.Split("|").ToList
        If fields.Count > fieldIdx Then
            retval = fields(fieldIdx)
        End If
        If retval = "" And alternativeIdx > 0 Then retval = Field(segment, alternativeIdx)
        Return retval
    End Function
End Class

File: matiasmasso/Mat.NET/DTO/DTOEdiversaInvoicD96a.cs
using MatHelperStd;
using System;
using System.Collections.Generic;
using System.Linq;

namespace DTO
{
    public class DTOEdiversaInvoicD96a : DTOBaseGuid
    {
        public string Inv { get; set; }
        public DateTime Dtm { get; set; }
        public string RffDq { get; set; }
        public string RffOn { get; set; }
        public string RffVn { get; set; }
        public string NadSu { get; set; }
        public string NadBy { get; set; }
        public string NadIv { get; set; }
        public string NadDp { get; set; }
        public string Cur { get; set; }
        public decimal MoaresNet { get; set; }
        public decimal MoaresBrut { get; set; }
        public decimal MoaresBase { get; set; }
        public decimal MoaresLiq { get; set; }
        public decimal MoaresTax { get; set; }
        public decimal MoaresDto { get; set; }
        public decimal MoaresCharges { get; set; }
        public List<Item> Items { get; set; }

        public class Item : DTOBaseGuid
        {
            public int Lin { get; set; }
            public string Ean { get; set; }
            public string PiaLin { get; set; }
            public string ImdLin { get; set; }
            public int QtyLin { get; set; }
            public decimal PriLinNet { get; set; }
            public decimal PriLinBrut { get; set; }
            public decimal MoaLinNet { get; set; }

            public Item() : base()
            {
            }

            public Item(Guid oGuid) : base(oGuid)
            {
            }
        }

        public DTOEdiversaInvoicD96a() : base()
        {
            Items = new List<Item>();
        }

        public DTOEdiversaInvoicD96a(Guid oGuid) : base(oGuid)
        {
            Items = new List<Item>();
        }

        public static DTOEdiversaInvoicD96a Factory(string src)
        {
            var segments = src.Split(new String[] { "\r\n" }, System.StringSplitOptions.RemoveEmptyEntries);
            DTOEdiversaInvoicD96a retval = new DTOEdiversaInvoicD96a();
            foreach (var segment in segments)
            {
                var fields = segment.Split('|');
                Item item = new Item();
                if (fields.Length > 1)
                {
                    switch (fields.First())
                    {
                        case "INV":
                            {
                                retval.Inv = Field(segment, 1);
                                break;
                            }

                        case "DTM":
                            {
                                retval.Dtm = ParseFch(segment, 1);
                                break;
                            }

                        case "RFF":
                            {
                                switch (Field(segment, 1))
                                {
                                    case "DQ":
                                        {
                                            retval.RffDq = Field(segment, 2);
                                            break;
                                        }

                                    case "ON":
                                        {
                                            retval.RffOn = Field(segment, 2);
                                            break;
                                        }

                                    case "VN":
                                        {
                                            retval.RffVn = Field(segment, 2);
                                            break;
                                        }
                                }

                                break;
                            }

                        case "NADSU":
                            {
                                retval.NadSu = Field(segment, 2);
                                break;
                            }

                        case "NADBY":
                            {
                                retval.NadBy = Field(segment, 2);
                                break;
                            }

                        case "NADIV":
                            {
                                retval.NadIv = Field(segment, 2);
                                break;
                            }

                        case "NADDP":
                            {
                                retval.NadDp = Field(segment, 2);
                                break;
                            }

                        case "CUX":
                            {
                                retval.Cur = Field(segment, 2);
                                break;
                            }

                        case "ALC":
                            {
                                break;
                            }

                        case "MOARES":
                            {
                                retval.MoaresNet = parseDecimal(segment, 1);
                                retval.MoaresBrut = parseDecimal(segment, 2);
                                retval.MoaresBase = parseDecimal(segment, 3);
                                retval.MoaresLiq = parseDecimal(segment, 4, 10);
                                retval.MoaresTax = parseDecimal(segment, 5);
                                retval.MoaresDto = parseDecimal(segment, 6, 13);
                                retval.MoaresCharges = parseDecimal(segment, 9);
                                break;
                            }

                        case "LIN":
                            {
                                item = new Item();
                                item.Ean = Field(segment, 1);
                                item.Lin = parseInt(segment, 3);
                                break;
                            }

                        case "PIALIN":
                            {
                                item.PiaLin = Field(segment, 1);
                                break;
                            }

                        case "IMDLIN":
                            {
                                item.ImdLin = Field(segment, 1);
                                break;
                            }

                        case "QTYLIN":
                            {
                                item.QtyLin = parseInt(segment, 2);
                                break;
                            }

                        case "MOALIN":
                            {
                                item.MoaLinNet = parseDecimal(segment, 3);
                                break;
                            }

                        case "PRILIN":
                            {
                                switch (Field(segment, 2))
                                {
                                    case "AAA":
                                        {
                                            item.PriLinNet = parseDecimal(segment, 2);
                                            break;
                                        }

                                    case "AAB":
                                        {
                                            item.PriLinBrut = parseDecimal(segment, 2);
                                            break;
                                        }
                                }

                                break;
                            }
                    }
                }
            }
            return retval;
        }

        public static int parseInt(string segment, int fieldIdx, int alternativeIdx = 0)
        {
            int retval = 0;
            string value = Field(segment, fieldIdx, alternativeIdx);
            if (TextHelper.VbIsNumeric(value))
                retval = value.toInteger();
            return retval;
        }

        public static decimal parseDecimal(string segment, int fieldIdx, int alternativeIdx = 0)
        {
            decimal retval = 0;
            string value = Field(segment, fieldIdx, alternativeIdx);
            try
            {
                retval = decimal.Parse(value, System.Globalization.CultureInfo.InvariantCulture);
            }
            catch (Exception)
            {
            }
            return retval;
        }

        public static DateTime ParseFch(string segment, int fieldIdx, int alternativeIdx = 0)
        {
            DateTime retval = default(DateTime);
            string value = Field(segment, fieldIdx, alternativeIdx);
            System.Globalization.CultureInfo provider = System.Globalization.CultureInfo.InvariantCulture;
            try
            {
                switch (value.Length)
                {
                    case 0:
                        {
                            break;
                        }

                    case 6:
                        {
                            retval = DateTime.ParseExact(value, "yyMMdd", provider);
                            break;
                        }

                    case 8:
                        {
                            retval = DateTime.ParseExact(value, "yyyyMMdd", provider);
                            break;
                        }

                    case 12:
                        {
                            retval = DateTime.ParseExact(value, "yyyyMMddHHmm", provider);
                            break;
                        }

                    default:
                        {
                            break;
                        }
                }
            }
            catch (Exception)
            {
            }

            return retval;
        }

        public static string Field(string segment, int fieldIdx, int alternativeIdx = 0)
        {
            string retval = "";
            var fields = segment.Split('|').ToList();
            if (fields.Count > fieldIdx)
                retval = fields[fieldIdx];
            if (retval == "" & alternativeIdx > 0)
                retval = Field(segment, alternativeIdx);
            return retval;
        }
    }
}
